worlds/pokemon_crystal/rules.py

```diff
@@ -0,0 +1,702 @@
+from typing import TYPE_CHECKING
+
+from BaseClasses import CollectionState
+from worlds.generic.Rules import add_rule, set_rule
+from .data import data
+from .options import JohtoOnly
+
+if TYPE_CHECKING:
+    from . import PokemonCrystalWorld
+
+
+def can_map_card_fly(state: CollectionState, world: "PokemonCrystalWorld"):
+    if world.options.randomize_pokegear:
+        return state.has("Map Card", world.player) and state.has("Pokegear", world.player)
+    return state.has("EVENT_GOT_MAP_CARD", world.player) and state.has("EVENT_GOT_POKEGEAR", world.player)
+
+
+def set_rules(world: "PokemonCrystalWorld") -> None:
+    if world.options.hm_badge_requirements == 0:
+        def can_cut(state: CollectionState):
+            return state.has("HM01 Cut", world.player) and has_badge(state, "hive")
+
+        def can_fly(state: CollectionState):
+            return state.has("HM02 Fly", world.player) and has_badge(state, "storm")
+
+        def can_surf(state: CollectionState):
+            return state.has("HM03 Surf", world.player) and has_badge(state, "fog")
+
+        def can_strength(state: CollectionState):
+            return state.has("HM04 Strength", world.player) and has_badge(state, "plain")
+
+        def can_flash(state: CollectionState):
+            return state.has("HM05 Flash", world.player) and has_badge(state, "zephyr")
+
+        def can_whirlpool(state: CollectionState):
+            return state.has("HM06 Whirlpool", world.player) and has_badge(state, "glacier") and can_surf(state)
```

Comment:
Is being able to surf truly a requirement for whirlpool and waterfall? Or is it only implicit from the fact that you're already on water when you encounter the roadblock.

In Emerald, I always have separate boundaries for water, waterfalls, and dive spots. This can separate encounter tables for dexsanity, so I didn't have to split more regions when I added it to account for water sources at the base of a waterfall. And there are some transitions that will put you on water automatically, which would allow you to use Waterfall/Dive without having Surf.

It wouldn't amount to much of a change in performance because of short circuiting, but I don't actually know if Crystal has warps that put you on water regardless of whether you have surf, and whether those connect to whirlpools or waterfalls. Or whether you plan on ever touching warp randomization.

Reply:
I can't think of anywhere in the game where a warp can be accessed while surfing, every door has some land in front of it to my knowledge. I will need to change how I do regions and rules for dexsanity and ER but those aren't in my immediate plans.